File: haskell/y2023/d01/Y2023D01.hs
module Main where

import Data.Char (isDigit)
import Data.List (isPrefixOf, isSuffixOf)
import Data.Map (Map, fromList, keys, (!))
import Input (readLines)

main :: IO ()
main = do
  lines <- readLines 2023 1
  p1 <- part1 lines
  putStrLn ("Part 1: " ++ show p1)
  p2 <- part2 lines
  putStrLn ("Part 2: " ++ show p2)

part1 :: [String] -> IO Int
part1 lines = do
  let nums = map (toNumber . digits) lines
  return (sum nums)

digits :: String -> String
digits = filter isDigit

toNumber :: String -> Int
toNumber s = read (head s : [last s])

part2 :: [String] -> IO Int
part2 lines = do
  let nums = map (\line -> startingNumber line * 10 + endingNumber line) lines
  return (sum nums)

numbers :: Map String Int
numbers = fromList [("one", 1), ("two", 2), ("three", 3), ("four", 4), ("five", 5), ("six", 6), ("seven", 7), ("eight", 8), ("nine", 9)]

startingNumber :: String -> Int
startingNumber s
  | isDigit (head s) = read [head s]
  | null matches = startingNumber (tail s)
  | otherwise = numbers ! head matches
  where
    matches = filter (`isPrefixOf` s) (keys numbers)

endingNumber :: String -> Int
endingNumber s
  | isDigit (last s) = read [last s]
  | null matches = endingNumber (init s)
  | otherwise = numbers ! head matches
  where
    matches = filter (`isSuffixOf` s) (keys numbers)

File: prolog/y2023/d01/y2023d01.pl
:- include('../../lib/input.pl').

main :-
    read_lines(Lines),
    part1(Lines, P1),
    write("Part 1: "), write(P1), nl,
    part2(Lines, P2),
    write("Part 2: "), write(P2), nl.

% Part 1

part1([], 0).
part1([Line|Lines], Result) :-
    findFirstNumber(Line, FirstNum),
    findLastNumber(Line, LastNum),
    part1(Lines, Results),
    Result is FirstNum * 10 + LastNum + Results.

findFirstNumber(S, Number) :-
    findall(Num, (sub_string(S, 0, _, _, Sub), number(Sub, Num)), [Number]).
findFirstNumber(S, Number) :-
    findall(Num, (sub_string(S, 0, _, _, Sub), number(Sub, Num)), []),
    sub_string(S, 1, _, 0, Rest),
    findFirstNumber(Rest, Number).

findLastNumber(S, Number) :-
    findall(Num, (sub_string(S, _, _, 0, Sub), number(Sub, Num)), [Number]).
findLastNumber(S, Number) :-
    findall(Num, (sub_string(S, _, _, 0, Sub), number(Sub, Num)), []),
    sub_string(S, 0, _, 1, Rest),
    findLastNumber(Rest, Number).

number("0", 0).
number("1", 1).
number("2", 2).
number("3", 3).
number("4", 4).
number("5", 5).
number("6", 6).
number("7", 7).
number("8", 8).
number("9", 9).

% Part 2

part2([], 0).
part2([Line|Lines], Result) :-
    findFirstWordNumber(Line, FirstNum),
    findLastWordNumber(Line, LastNum),
    part2(Lines, Results),
    Result is FirstNum * 10 + LastNum + Results.

findFirstWordNumber(S, Number) :-
    findall(Num, (sub_string(S, 0, _, _, Sub), wordNumber(Sub, Num)), [Number]).
findFirstWordNumber(S, Number) :-
    findall(Num, (sub_string(S, 0, _, _, Sub), wordNumber(Sub, Num)), []),
    sub_string(S, 1, _, 0, Rest),
    findFirstWordNumber(Rest, Number).

findLastWordNumber(S, Number) :-
    findall(Num, (sub_string(S, _, _, 0, Sub), wordNumber(Sub, Num)), [Number]).
findLastWordNumber(S, Number) :-
    findall(Num, (sub_string(S, _, _, 0, Sub), wordNumber(Sub, Num)), []),
    sub_string(S, 0, _, 1, Rest),
    findLastWordNumber(Rest, Number).

wordNumber("zero", 0).
wordNumber("one", 1).
wordNumber("two", 2).
wordNumber("three", 3).
wordNumber("four", 4).
wordNumber("five", 5).
wordNumber("six", 6).
wordNumber("seven", 7).
wordNumber("eight", 8).
wordNumber("nine", 9).
wordNumber(Word, Num) :- 
    number(Word, Num).
